### ISSUE TITLE
AxonDeepSeg is not compatible with Windows

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/axondeepseg#91]
What are the limitations and how can we address them?

**Comment by mathieuboudreau** [comment]
Just created a fresh Windows 10 VirtualBox VM and tested it.

- Installed Python 2.7 using the [Anaconda package](https://www.anaconda.com/download/).
- Opened up an Anaconda path session through the `Anaconda Prompt` logo in the App list
    - Had to run it using Administrator privileges to be able to install packages later on.
- Installed pip in using `conda install pip` on the command line.
- Installed git in using `conda install git`on the command line.
- Cloned ADS using the command line and CD'd into it.
- Created a virtual environment using `conda create --name ads_venv`
- Activated it using `conda activate ads_venv`
- Installed the ADS dependencies using `pip install -e .`

No problems yet.

- Run the installation test `axondeepseg_test`, fails, see screenshot below:

![ads](https://user-images.githubusercontent.com/1421029/39585003-5518b5d0-4ece-11e8-99a6-31d494abd2ac.png)

This was my first time running Python in windows, so I don't have any further insights yet on why this might have been thrown considering that all the dependencies seem to have installed fine.

**Comment by mathieuboudreau** [comment]
See: https://stackoverflow.com/questions/3571514/python-why-should-from-module-import-be-prohibited

Not sure why this would fail in Windows but pass in OS/UNIX, I might try importing the modules outside the function definition: https://github.com/neuropoly/axondeepseg/blob/bfb0fe3ae65c97671df2aa01bbbc147135d7fc21/AxonDeepSeg/integrity_test.py#L1-L16

**Comment by mathieuboudreau** [comment]
Hmmmm actually I just double checked, even though I specifically installed Anaconda 2.7, `python -V` tells me I'm using Python 3.6.5, so that might be the source of that error. I'll debug that issue and try the test again.

**Comment by perone** [comment]
I don't see any limitation that would block the use of it on Windows, I think all required packages are compatible. What can happen is some issues with paths/etc, but it requires testing as @mathieuboudreau is doing, it's the only way to fix the issues that pop out.

**Comment by mathieuboudreau** [comment]
**Looks like using Windows + Python 2.7 + Tensorflow is a (the?) major limitation preventing us from Windows compatibility.**

- I explicitly installed a Python 2.7 venv (`conda create -n ads_venv27 python=2.7) and activated it.
- Ran `pip install -e .` and encountered an error saying that the required TensorFlow version couldn't be met:

![ads27](https://user-images.githubusercontent.com/1421029/39586123-ef93da52-4ed0-11e8-8691-06f5991eedf8.png)

Googled "tensorflow windows" and found [this official TensorFlow documentation page](https://www.tensorflow.org/install/install_windows#installing_with_native_pip), which states that Python 3.5x or 3.6x are required to use TensorFlow on windows.

[This stackoverflow post](https://stackoverflow.com/questions/45316569/how-to-install-tensorflow-on-python-2-7-on-windows) also seems to suggest that you must be running 3.5x/3.6x on Windows to use Tensorflow.

**Comment by alzaia** [comment]
Interesting... Thanks a lot for the investigation @mathieuboudreau ! So it looks like we would need to migrate to 3.5x or 3.6x to make it run on Windows.

**Comment by jcohenadad** [comment]
awesome investigations @mathieuboudreau!!!  Is the current version of ADS happy with Python 3.5? If not, maybe we should open a specific issue for making it compatible, what do you think?

**Comment by mathieuboudreau** [comment]
Thanks @jcohenadad !

I just tried doing the necessary steps to convert the code to be Python 3 compatible, and early indications suggest that it shouldn't be too difficult, with a few caveats. 

Working on branch: [py2to3](https://github.com/neuropoly/axondeepseg/tree/py2to3)

First-order test was getting `axondeepseg_test` to pass in Python 3:

- Converted the repo's code using the command [2to3](https://docs.python.org/2/library/2to3.html)
    - All the conversions did not pass. Received the following warnings:
```
RefactoringTool: Warnings/messages while refactoring:
RefactoringTool: ### In file ./AxonDeepSeg/config_tools.py ###
RefactoringTool: Line 226: could not convert: raise 'Invalid input : please for features_augmentation'
RefactoringTool: Python 3 does not support string exceptions
RefactoringTool: ### In file ./AxonDeepSeg/data_management/input_data.py ###
RefactoringTool: Line 41: You should use a for loop here
RefactoringTool: There was 1 error:
RefactoringTool: Can't parse ./AxonDeepSeg/visualization/generate_axons_from_myelin.py: IndentationError: unindent does not match any outer indentation level (<tokenize>, line 22)
```
but moved ahead without fixing these manually yet.
- Moved the module imports outside of a function in integrity_test.py – seems like this isn't compatible in Python 3, as it threw an error otherwise. Haven't browsed through the repos code to see if it occurs anywhere else.
- Converted certain function inputs into `int` type in several locations of the function `uconv_net()` of network_construction.py, as it was complaining that Tensorflow was expecting int's but getting floats. Either Tensorflow in 2.7 environments converted these values implicitely, or uconv_net() did so, but this seems to be a bug that `2to3` didn't know to resolve. Here's one of the errors before the fix:

```
(ads_venv36) mathieuboudreau@brassens:~/neuropoly/github/axondeepseg$ axondeepseg_test 
/Users/mathieuboudreau/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Reading test configuration file.
Computing the segmentation of axon and myelin on test image.
Tensor("downconv-d3/convolution/Relu:0", shape=(?, 32, 32, 128), dtype=float32)
64.0
64.0
[64.0, 64.0]
Traceback (most recent call last):
  File "/Users/mathieuboudreau/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/image_ops_impl.py", line 907, in resize_images
    size = ops.convert_to_tensor(size, dtypes.int32, name='size')
  File "/Users/mathieuboudreau/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1014, in convert_to_tensor
    as_ref=False)
  File "/Users/mathieuboudreau/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1104, in internal_convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/Users/mathieuboudreau/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/constant_op.py", line 235, in _constant_tensor_conversion_function
    return constant(v, dtype=dtype, name=name)
  File "/Users/mathieuboudreau/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/constant_op.py", line 214, in constant
    value, dtype=dtype, shape=shape, verify_shape=verify_shape))
  File "/Users/mathieuboudreau/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/tensor_util.py", line 432, in make_tensor_proto
    _AssertCompatible(values, dtype)
  File "/Users/mathieuboudreau/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/tensor_util.py", line 343, in _AssertCompatible
    (dtype.name, repr(mismatch), type(mismatch).__name__))
TypeError: Expected int32, got 64.0 of type 'float' instead.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mathieuboudreau/anaconda3/bin/axondeepseg_test", line 11, in <module>
    load_entry_point('AxonDeepSeg', 'console_scripts', 'axondeepseg_test')()
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/integrity_test.py", line 39, in integrity_test
    prediction = axon_segmentation([path_testing], ["image.png"], path_model, config_network,verbosity_level=0)
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/apply_model.py", line 260, in axon_segmentation
    gpu_per=gpu_per, verbosity_level=verbosity_level)
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/apply_model.py", line 67, in apply_convnet
    pred = uconv_net(x, config_dict, phase=False, verbose=False)  # Inference
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/network_construction.py", line 200, in uconv_net
    net = tf.image.resize_images(data_temp, [data_temp_size[-1] * 2, data_temp_size[-1] * 2])
  File "/Users/mathieuboudreau/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/image_ops_impl.py", line 909, in resize_images
    raise ValueError('\'size\' must be a 1-D int32 Tensor')
ValueError: 'size' must be a 1-D int32 Tensor
```

After these three steps, `axondeepseg_test` passes in Python 3 on my Mac:

![mac](https://user-images.githubusercontent.com/1421029/39631934-37508f80-4f8e-11e8-92c4-3783d27f759c.png)

**and on Windows**:

![windows](https://user-images.githubusercontent.com/1421029/39631992-62b1b942-4f8e-11e8-8f1e-365b2a53ba26.png)

However, since `axondeepseg_test` only has partial code coverage of the repo, I suspect there are several other fixes needed, and I'm not experienced enough with this project yet to know how to test all of it (e.g. generating new models, testing all the options, etc).

**Comment by jcohenadad** [comment]
Great contribution @mathieuboudreau! thanks a lot. @alzaia can you please help Mathieu as you are the most familiar with the project.

**Comment by mathieuboudreau** [comment]
Resolved in release v2.0.
